spec.emu

```diff
@@ -835,12 +1711,179 @@ contributors: Luca Casonato, Guy Bedford, Nicolò Ribaudo
+        <li>has the following properties:</li>
+      </ul>
+
+      <emu-clause id="sec-%typedarray%.prototype">
+        <h1>%AbstractModuleSource%.prototype</h1>
+        <p>The initial value of %AbstractModuleSource%`.prototype` is the %AbstractModuleSource% prototype object.</p>
+        <p>This property has the attributes { [[Writable]]: *false*, [[Enumerable]]: *false*, [[Configurable]]: *false* }.</p>
+      </emu-clause>
+    </emu-clause>
+
+    <emu-clause id="sec-properties-of-the-%abstractmodulesource%-prototype-object">
+      <h1>Properties of the %AbstractModuleSource% Prototype Object</h1>
+      <p>The <dfn>%AbstractModuleSource% prototype object</dfn>:</p>
+      <ul>
+        <li>has a [[Prototype]] internal slot whose value is %Object.prototype%.</li>
+        <li>is <dfn>%AbstractModuleSource.prototype%</dfn>.</li>
+        <li>is an ordinary object.</li>
+        <li>does not have a global name or appear as a property of the global object.</li>
+        <li>does not have a [[ModuleSourceRecord]] slot that is specific to _AbstractModuleSource_ and _AbstractModuleSource_ subclass instance objects.</li>
+        <li>has the following properties:</li>
+      </ul>
+
+      <emu-clause id="sec-%abstractmodulesource%.prototype.constructor">
+        <h1>%AbstractModuleSource%.prototype.constructor</h1>
+        <p>The initial value of %AbstractModuleSource%`.prototype.constructor` is %AbstractModuleSource%.</p>
+      </emu-clause>
+
+      <emu-clause id="sec-get-%abstractmodulesource%.prototype.@@tostringtag">
+        <h1>get %AbstractModuleSource%.prototype [ @@toStringTag ]</h1>
+        <p>%AbstractModuleSource%.prototype `[@@toStringTag]` is an accessor property whose set accessor function is *undefined*. Its get accessor function performs the following steps when called:</p>
+        <emu-alg>
+          1. Let _O_ be the *this* value.
+          1. If _O_ is not an Object, return *undefined*.
+          1. If _O_ does not have a [[ModuleSourceRecord]] internal slot, return *undefined*.
+          1. Let _name_ be _O_.[[ModuleSourceRecord]].[[SourceClassName]]
+          1. Assert: _name_ is a String.
+          1. Return _name_.
+        </emu-alg>
+        <p>This property has the attributes { [[Enumerable]]: *false*, [[Configurable]]: *true* }.</p>
+        <p>The initial value of the *"name"* property of this function is *"get [Symbol.toStringTag]"*.</p>
+      </emu-clause>
+    </emu-clause>
+
+    <emu-clause id="sec-modulesource-constructor">
+      <h1>The ModuleSource Constructor</h1>
+      <p>The ModuleSource Object represents the source phase of a SourceTextModuleRecord.</p>
+
+      <p>The ModuleSource constructor:</p>
+      <ul>
+        <li>is <dfn>%ModuleSource%</dfn>.</li>
+        <li>is the initial value of the *"ModuleSource"* property of the global object.</li>
+        <li>has a [[Prototype]] internal slot whose value is %AbstractModuleSource.prototype%.</li>
```

Comment:
```suggestion
        <li>has a [[Prototype]] internal slot whose value is %AbstractModuleSource%.</li>
```

```js
class A {}
class B extends A {}
B.__proto__ === A;
```

Reply:
I realize now that this is a good reason to have `%AbstractModuleSource%`. All the case I listed in a comment above where the constructor doesn't exist are never extended by an existing constructor.

Reply:
//cc @kriskowal do you have any strong opinions here with regards to retaining an `AbstractModuleSource` constructor explicitly existing.

Reply:
I don’t have any strong opinions about these details.

Reply:
Okay, so this is a strong argument for the %AbstractModuleSource% constructor as the shared home of static methods for module sources per standard classes in JS. Agreed this model works out at this point, so that seems to settle the `AbstractModuleSource` structuring question.